src/components/InngestStepTools.ts

```diff
@@ -88,6 +89,20 @@ export const createStepTools = <
      * tick has completed.
      */
     reset: () => void;
+
+    /**
+     * If `true`, any use of step tools will, by default, throw an error. We do
+     * this when we detect that a function may be mixing step and non-step code.
+     *
+     * Created step tooling can decide how to manually handle this on a
+     * case-by-case basis.
+     *
+     * In the future, we can provide a way for a user to override this if they
+     * wish to and understand the danger of side-effects.
+     *
+     * Defaults to `false`.
+     */
+    nonStepFnDetected: boolean;
```

Comment:
Can this be computed instead of making it a flag?

Reply:
@darwin67 We'd still need to know a little more about the state than we currently do in order to calculate this on-the-fly, at least that we've done memoizing state.

That in itself would require a flag or us marking stack items as we pass them, so might not be worth it here.

Agreed, though - there are a probably a few other places this approach would help clear up logic!